### PR TITLE
Feature/rb 54 collision system

### PR DESCRIPTION
**Pull request by KitetsuK**
* **Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)
- [x] Norm of the code has been respected

* **In what state is your pull request?**
- [x] Ready to merge / Waiting a reviwer to see my work.
- [ ] Work In Progress (WIP) / My work is not finish but i want daily reviews. (Draft)
- [ ] CI Review / I want to see what the CI thinks of my work.
- [ ] Need feedback / Just to have feedback on what i produced. (May not be merged)

* **What kind of change does this PR introduce?** (You can choose multiple)
- [ ] Bug fix
- [x] Feature request
- [ ] New / Updated documentation
- [ ] Testing CI ( Make the pull request in draft mode)

* **What is the current behavior?** (link an issue based on the kind of change this pr introduce)

* **What is the new behavior (if this is a feature change)?**
Add a collision system between entities
An entity can collide multiple entities
Damage dealing is implemented inside the collision system for now but we re looking forward to move it inside a system

* **Other information**:
New custom types had been created

### CODE REVIEW

File: src/ECS/Systems/Systems.cpp [hunk 1]
```diff
@@ -48,8 +48,91 @@ namespace Systems {
...
+            Registry::getInstance().getComponents<Types::Position>();
+        Registry::components<Types::CollisionRect> arrCollisionRect =
+            Registry::getInstance().getComponents<Types::CollisionRect>();
+
+        auto collisionIt = arrCollisionRect.begin();
+        auto positionIt  = arrPosition.begin();
+
+        while (collisionIt != arrCollisionRect.end()
+               && positionIt != arrPosition.end()) {
+            if (collisionIt->has_value() && positionIt->has_value()) {
```
Comment (kw4ntiq) [inline src/ECS/Systems/Systems.cpp]: sachant que collisionIt++ et positionIt++ sont dans le "if (collisionIt->has_value() && positionIt->has_value())"
Si une entite n'a pas de collision et ou de position ca va boucle inf non ?

Comment (KitetsuK) [reply src/ECS/Systems/Systems.cpp]: Bien vu je fix ça demain

File: src/ECS/Systems/Systems.cpp [hunk 2]
```diff
@@ -48,8 +48,91 @@ namespace Systems {
...
+                        > tmpPositionIt->value().y) {
+                    giveDamages(
+                        std::distance(arrPosition.begin(), positionIt),
+                        std::distance(arrPosition.begin(), tmpPositionIt));
+                    tmpCollisionIt++;
+                    tmpPositionIt++;
+                } else {
+                    tmpCollisionIt++;
+                    tmpPositionIt++;
+                }
```
Comment (guillaumeAbel) [inline src/ECS/Systems/Systems.cpp]:
```suggestion
                }
            }
```

File: src/ECS/Systems/Systems.cpp [hunk 3]
```diff
@@ -48,8 +48,91 @@ namespace Systems {
...
+
+    static void checkCollisionEntity(
+        std::vector<std::optional<Types::CollisionRect>>::iterator collisionIt,
+        std::vector<std::optional<Types::Position>>::iterator positionIt,
+        Registry::components<Types::Position> arrPosition,
+        Registry::components<Types::CollisionRect> arrCollision
+
+    )
+    {
+        auto tmpCollisionIt = collisionIt;
```
Comment (guillaumeAbel) [inline src/ECS/Systems/Systems.cpp]: collisionIt est deja une copie de l'iterateur passé en parametre vu que c'est pas une ref ou un ptr

File: src/ECS/Systems/Systems.cpp [hunk 4]
```diff
@@ -48,8 +48,91 @@ namespace Systems {
...
+        auto tmpPositionIt  = positionIt;
+
+        tmpPositionIt++;
+        tmpCollisionIt++;
```
Comment (guillaumeAbel) [inline src/ECS/Systems/Systems.cpp]:
```suggestion
```

File: src/ECS/Systems/Systems.cpp [hunk 5]
```diff
@@ -48,8 +48,91 @@ namespace Systems {
...
+        while (tmpCollisionIt != arrCollision.end()
+               && tmpPositionIt != arrPosition.end()) {
+            if (tmpCollisionIt->has_value() && tmpPositionIt->has_value()) {
```
Comment (guillaumeAbel) [inline src/ECS/Systems/Systems.cpp]:
```suggestion
            collisionIt++
            positionIt++
            if (tmpCollisionIt->has_value() && tmpPositionIt->has_value()) {
```

File: src/ECS/Systems/Systems.cpp [hunk 6]
```diff
@@ -48,8 +48,91 @@ namespace Systems {
         }
     }
 
+    static void giveDamages(std::size_t firstEntity, std::size_t secondEntity)
+    {
+        std::optional<Types::Health> &firstEntityHealth =
```
Comment (guillaumeAbel) [inline src/ECS/Systems/Systems.cpp]: code duplicata, call 2 times a function doing the if

**Review by guillaumeAbel — CHANGES_REQUESTED** [review]

File: src/ECS/Systems/Systems.cpp [hunk 7]
```diff
@@ -48,8 +48,91 @@ namespace Systems {
...
+            Registry::getInstance()
+                .getComponents<Types::Dammage>()[firstEntity];
+        std::optional<Types::Health> &secondEntityHealth =
+            Registry::getInstance()
+                .getComponents<Types::Health>()[secondEntity];
+        std::optional<Types::Dammage> &secondEntityDammage =
+            Registry::getInstance()
+                .getComponents<Types::Dammage>()[secondEntity];
+
+        if (firstEntityDammage.has_value() && secondEntityHealth.has_value()) {
```
Comment (guillaumeAbel) [inline src/ECS/Systems/Systems.cpp]: u dont check infinite health or no dammage

**Review by guillaumeAbel — CHANGES_REQUESTED** [review]

**Review by Saverio976 — APPROVED** [review]

**Review by guillaumeAbel — APPROVED** [review]

**Pull request merged**